components/layout/display_list_builder.rs

```diff
@@ -608,57 +608,81 @@ fn convert_gradient_stops(gradient_items: &[GradientItem],
     // Determine the position of each stop per CSS-IMAGES § 3.4.
     //
     // FIXME(#3908, pcwalton): Make sure later stops can't be behind earlier stops.
-    let stop_items = gradient_items.iter().filter_map(|item| {
+
+    // Only keep the color stops. Discard the color items.
+    let mut stop_items = gradient_items.iter().filter_map(|item| {
         match *item {
-            GradientItem::ColorStop(ref stop) => Some(stop),
+            GradientItem::ColorStop(ref stop) => Some(*stop),
             _ => None,
         }
     }).collect::<Vec<_>>();
-    let mut stops = Vec::with_capacity(stop_items.len());
+
+    assert!(stop_items.len() >= 2);
+
+    // Run the algorithm from
+    // https://drafts.csswg.org/css-images-3/#color-stop-syntax
+
+    // Step 1: If nothing else is specified the first stop is
+    // at 0% and the last stop is at 100%.
+    {
+        let first = stop_items.first_mut().unwrap();
+        if first.position.is_none() {
+            first.position = Some(LengthOrPercentage::Percentage(0.0));
+        }
+    }
+    {
+        let last = stop_items.last_mut().unwrap();
+        if last.position.is_none() {
+            last.position = Some(LengthOrPercentage::Percentage(1.0));
+        }
+    }
+
+    // Step 2: Move any stops placed before earlier stops to the
+    // same position as the preceding stop.
+    let mut state = stop_items.first().unwrap().position.unwrap();
```

Comment:
instead of `state`, what about `last_stop_position`?

Also, we could move it to its own function to make this one a bit less dense, but as you prefer.